File: backend/app/core/password_validator.py
"""
Password Complexity Validator

Centralised validation rules for all password creation / change flows.
Returns a list of human-readable violation messages (empty list = valid).

Rules:
  1. Minimum 8 characters
  2. At least one uppercase letter
  3. At least one digit
  4. At least one special character
  5. Must not contain the email local part (before @)
  6. Must not be in a blocklist of commonly guessable passwords
"""

import re

# Passwords that trivially satisfy all character-class rules but are still
# highly guessable.  Keep this list small and targeted.
COMMON_PASSWORD_BLOCKLIST = {
    "Password123!",
    "Qwerty123!",
    "Admin123!",
    "Welcome1!",
    "Changeme1!",
    "Abcd1234!",
    "P@ssw0rd!",
    "Passw0rd!",
    "Test1234!",
    "Letmein1!",
}

_SPECIAL_CHARS = re.compile(r"[!@#$%^&*()\-_+=\[\]{}|;:'\",.<>?/`~\\]")


def validate_password_complexity(password: str, email: str = "") -> list[str]:
    """
    Validate a password against the project's complexity policy.

    Args:
        password: The candidate password string.
        email:    The user's email (used to check that the password does not
                  contain the local part of the email).

    Returns:
        A list of violation messages.  An empty list means the password is valid.
    """
    errors: list[str] = []

    if len(password) < 8:
        errors.append("Password must be at least 8 characters long.")

    if not any(c.isupper() for c in password):
        errors.append("Password must contain at least one uppercase letter.")

    if not any(c.isdigit() for c in password):
        errors.append("Password must contain at least one digit.")

    if not _SPECIAL_CHARS.search(password):
        errors.append("Password must contain at least one special character.")

    # Check if the password contains the email local-part (case-insensitive)
    if email:
        local_part = email.split("@")[0].lower()
        if len(local_part) >= 3 and local_part in password.lower():
            errors.append("Password must not contain your email username.")

    # Blocklist check (case-sensitive — intentional; these are exact patterns)
    if password in COMMON_PASSWORD_BLOCKLIST:
        errors.append("This password is too common. Please choose a stronger one.")

    return errors

File: frontend/src/utils/passwordValidator.ts
/**
 * Password Complexity Validator (client-side mirror)
 *
 * Mirrors the backend rules in core/password_validator.py so the UI can
 * provide real-time feedback without a round-trip to the server.
 */

export interface PasswordRequirement {
  /** Human-readable description of the requirement. */
  label: string;
  /** Whether the current password satisfies this requirement. */
  met: boolean;
}

const SPECIAL_CHAR_REGEX = /[!@#$%^&*()\-_+=[\]{}|;:'",.<>?/`~\\]/;

const COMMON_BLOCKLIST = new Set([
  'Password123!',
  'Qwerty123!',
  'Admin123!',
  'Welcome1!',
  'Changeme1!',
  'Abcd1234!',
  'P@ssw0rd!',
  'Passw0rd!',
  'Test1234!',
  'Letmein1!',
]);

/**
 * Evaluate password against all complexity requirements.
 *
 * @param password - The candidate password.
 * @param email    - Optional user email. If provided, checks that the password
 *                   does not contain the email local-part.
 * @returns An array of requirement objects with `met` status.
 */
export function validatePassword(
  password: string,
  email?: string,
): PasswordRequirement[] {
  const requirements: PasswordRequirement[] = [
    {
      label: 'At least 8 characters',
      met: password.length >= 8,
    },
    {
      label: 'At least one uppercase letter',
      met: /[A-Z]/.test(password),
    },
    {
      label: 'At least one digit',
      met: /\d/.test(password),
    },
    {
      label: 'At least one special character',
      met: SPECIAL_CHAR_REGEX.test(password),
    },
    {
      label: 'Not a commonly used password',
      met: !COMMON_BLOCKLIST.has(password),
    },
  ];

  // Only add the email check when an email is provided and has a meaningful local part
  if (email) {
    const localPart = email.split('@')[0]?.toLowerCase() ?? '';
    if (localPart.length >= 3) {
      requirements.push({
        label: 'Must not contain your email username',
        met: !password.toLowerCase().includes(localPart),
      });
    }
  }

  return requirements;
}

/**
 * Returns true if all password requirements are met.
 */
export function isPasswordValid(password: string, email?: string): boolean {
  return validatePassword(password, email).every((r) => r.met);
}
